lib/kafka/ssl_socket_with_timeout.rb

```diff
@@ -0,0 +1,158 @@
+      @tcp_socket.setsockopt(Socket::IPPROTO_TCP, Socket::TCP_NODELAY, 1)
+
+      # first initiate the TCP socket
+      begin
+        # Initiate the socket connection in the background. If it doesn't fail 
+        # immediately it will raise an IO::WaitWritable (Errno::EINPROGRESS) 
+        # indicating the connection is in progress.
+        @tcp_socket.connect_nonblock(sockaddr)
+      rescue IO::WaitWritable
+        # IO.select will block until the socket is writable or the timeout
+        # is exceeded, whichever comes first.
+        unless IO.select(nil, [@tcp_socket], nil, connect_timeout)
+          # IO.select returns nil when the socket is not ready before timeout 
+          # seconds have elapsed
+          @tcp_socket.close
+          raise Errno::ETIMEDOUT
+        end
+
+        begin
+          # Verify there is now a good connection.
+          @tcp_socket.connect_nonblock(sockaddr)
+        rescue Errno::EISCONN
+          # The socket is connected, we're good!
+        end
+      end
+
+      # once that's connected, we can start initiating the ssl socket
+      if ssl_context
+        @ssl_socket = OpenSSL::SSL::SSLSocket.new(@tcp_socket, ssl_context)
+      else
+        @ssl_socket = OpenSSL::SSL::SSLSocket.new(@tcp_socket)
+      end
+
+      begin
+        # Initiate the socket connection in the background. If it doesn't fail 
+        # immediately it will raise an IO::WaitWritable (Errno::EINPROGRESS) 
+        # indicating the connection is in progress.
+        # Unlike waiting for a tcp socket to connect, you can't time out ssl socket
+        # connections during the connect phase properly, because IO.select only partially works.
+        # Instead, you have to retry.
+        @ssl_socket.connect_nonblock
+      rescue Errno::EAGAIN, Errno::EWOULDBLOCK, IO::WaitReadable
+        IO.select([@ssl_socket])
+        retry
+      rescue IO::WaitWritable
+        IO.select(nil, [@ssl_socket])
+        retry
+      end
+    end
+
+    # Reads bytes from the socket, possible with a timeout.
+    #
+    # @param num_bytes [Integer] the number of bytes to read.
+    # @raise [Errno::ETIMEDOUT] if the timeout is exceeded.
+    # @return [String] the data that was read from the socket.
+    def read(num_bytes)
+      buffer = ''
+      until buffer.length >= num_bytes
+        begin
+          # unlike plain tcp sockets, ssl sockets don't support IO.select
+          # properly.
+          # Instead, timeouts happen on a per read basis, and we have to
+          # catch exceptions from read_nonblock, and gradually build up
+          # our read buffer.
+          buffer << @ssl_socket.read_nonblock(num_bytes - buffer.length)
+        rescue IO::WaitReadable
+          unless IO.select([@ssl_socket], nil, nil, @timeout)
+            raise Errno::ETIMEDOUT
+          end
+          retry
+        rescue IO::WaitWritable
+          unless IO.select(nil, [@ssl_socket], nil, @timeout)
```

Comment:
it doesn't, and the SSLSocket docs are wrong as far as I can tell. I looked at two different ruby http libraries, and they all take this approach.

Your example doesn't work because the TCP socket can (and does) have data to read that's just for SSL communication, and nothing to do with the underlying layer.